### PR TITLE
Add initial implementation of the tracez page.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Debugging page that displays data about all the active Spans and samples based on latency and errors. Used to debug real-time problems.

tracez

### CODE REVIEW

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+    Formatter formatter = new Formatter(out, Locale.US);
+    emitSummaryTable(out, formatter);
+    String spanName = queryMap.get(HEADER_SPAN_NAME);
+    if (spanName != null) {
+      tracer
+          .getCurrentSpan()
+          .addAnnotation(
+              "Render spans.",
+              ImmutableMap.<String, AttributeValue>builder()
+                  .put("SpanName", AttributeValue.stringAttributeValue(spanName))
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Please escape all HTML output coming from user

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+                  .build());
+      String typeStr = queryMap.get(HEADER_SAMPLES_TYPE);
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Same, there are other places bellow

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+   * Emits the HTML generated page to the {@code outputStream}.
+   *
+   * @param queryMap the query components map.
+   * @param outputStream the output {@code OutputStream}.
+   */
+  public void emitHtml(Map<String, String> queryMap, OutputStream outputStream) {
+    PrintWriter out =
+        new PrintWriter(
+            new BufferedWriter(new OutputStreamWriter(outputStream, Charset.forName("US-ASCII"))));
+    out.write("<html><head><title>Tracez</title>");
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: You should also emit DOCTYPE declaration

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+              "Render spans.",
+              ImmutableMap.<String, AttributeValue>builder()
+                  .put("SpanName", AttributeValue.stringAttributeValue(spanName))
+                  .build());
+      String typeStr = queryMap.get(HEADER_SAMPLES_TYPE);
+      if (typeStr != null) {
+        List<SpanData> spans = null;
+        int type = Integer.parseInt(typeStr);
+        if (type < 0 || type > TYPE_ERROR) {
+          return;
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: this will stop HTML generation in bad way.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+   */
+  public void emitHtml(Map<String, String> queryMap, OutputStream outputStream) {
+    PrintWriter out =
+        new PrintWriter(
+            new BufferedWriter(new OutputStreamWriter(outputStream, Charset.forName("US-ASCII"))));
+    out.write("<html><head><title>Tracez</title>");
+    out.write("<link rel=\"shortcut icon\" href=\"//www.opencensus.io/favicon.ico\"/>");
+    out.write("</head>\n");
+    if (runningSpanStore == null || sampledSpanStore == null) {
+      out.write("OpenCensus implementation not available.");
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: This will cause bad HTML to be generated

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+                  runningSpanStore.getRunningSpans(RunningSpanStore.Filter.create(spanName, 0)));
+          // Sort active spans incremental.
+          Collections.sort(spans, new SpanDataComparator(true));
+        } else {
+          String subtypeStr = queryMap.get(HEADER_SAMPLES_SUB_TYPE);
+          if (subtypeStr != null) {
+            int subtype = Integer.parseInt(subtypeStr);
+            if (type == TYPE_ERROR) {
+              if (subtype < 0 || subtype >= CanonicalCode.values().length) {
+                return;
```
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Same. I would propose to put the "meat" of HTML generation into a separate method and throw in case of errors. The "parent" method will generate HTML header/tail and will also produce a nice error if/when exception is thrown.

File: contrib/zpages/README.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# OpenCensus Z-Pages
+[![Build Status][travis-image]][travis-url] [![Build status][appveyor-image]][appveyor-url] [![Maven Central][maven-image]][maven-url]
+
+The *OpenCensus ZPages for Java* is a collection of HTML pages to display stats and trace data and
```
Comment (bogdandrutu) [inline contrib/zpages/README.md]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 3 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 5 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 1 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: This is not an HTML output. This is where I am using the tracing library to create my own Span. Generated HTML code for this span will be escaped at line 323.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 2 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: What is the problem with this?

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 4 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 6 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 1 above)
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: You  are correct. However the renderAnnotation() doesn't escape the annotation, does it?

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 2 above)
Comment (AlekseySanin) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: The general attack vector is that you don't trust the data from user. Thus all the user data should be properly escaped when rendering html pages.

Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: We use this as an Integer so this should be safe. The only string in the components is spanName which when I use in href I do URLEncode (line 470) and when I print in the HTML page I do HTMLEncode (line 219).

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 1 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: The result of the renderAnnotation(...) is escaped:
```
          htmlEscaper()
              .escape(
                  event.getEvent() instanceof Annotation
                      ? renderAnnotation((Annotation) event.getEvent())
                      : renderNetworkEvents((NetworkEvent) event.getEvent());
```

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+   *
+   * @param server the server that exports the tracez page.
+   */
+  public static void register(HttpServer server) {
+    server.createContext(TRACEZ_URL, new TracezHttpHandler());
+  }
+
+  @Override
+  public final void handle(HttpExchange httpExchange) throws IOException {
+    try (Scope ss =
```
Comment (adriancole) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: why am I unsurprised to see tracez traced...

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        tracer
+            .spanBuilderWithExplicitParent(HTTP_SERVER_SPAN_NAME, null)
+            .setRecordEvents(true)
+            .startScopedSpan()) {
+      tracer
+          .getCurrentSpan()
+          .addAttributes(
```
Comment (adriancole) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: makes me think there should be an addAttribute method

File: settings.gradle [hunk 10]
```diff
@@ -12,20 +11,21 @@ include ":core"
...
 project(':opencensus-impl').projectDir = "$rootDir/impl" as File
 project(':opencensus-testing').projectDir = "$rootDir/testing" as File
 project(':opencensus-trace-logging-exporter').projectDir = "$rootDir/exporters/trace_logging" as File
+project(':opencensus-agent').projectDir = "$rootDir/contrib/agent" as File
 
 // Java8 projects only
 if (JavaVersion.current().isJava8Compatible()) {
     include ":examples"
     include ":benchmarks"
+    include ":opencensus-zpages"
```
Comment (adriancole) [inline settings.gradle]: for consistency call it zpages?

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+              ImmutableMap.<String, AttributeValue>builder()
+                  .put(
+                      "RequestMethod",
+                      AttributeValue.stringAttributeValue(httpExchange.getRequestMethod()))
+                  .build());
+      httpExchange.sendResponseHeaders(200, 0);
+      pageFormatter.emitHtml(queryToMap(httpExchange), httpExchange.getResponseBody());
+      httpExchange.close();
```
Comment (adriancole) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: maybe move close to a finally?

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+
+  private static Map<String, String> queryToMap(HttpExchange httpExchange) {
```
Comment (adriancole) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: easier to unit test if you make this package private

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    Map<String, String> result = new HashMap<String, String>();
+    for (String param : query.split("&")) {
+      String[] pair = param.split("=");
+      if (pair.length > 1) {
+        result.put(pair[0], pair[1]);
+      } else {
+        result.put(pair[0], "");
+      }
+    }
+    tracer.getCurrentSpan().addAnnotation("Finish to parse query components.");
```
Comment (adriancole) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: is this for debugging or do you expect to leave this in?

File: contrib/zpages/build.gradle [hunk 14]
```diff
@@ -0,0 +1,14 @@
+description = 'OpenCensus Z-Pages'
+
+apply plugin: 'java'
+
+[compileJava, compileTestJava].each() {
+    it.sourceCompatibility = 1.8
+    it.targetCompatibility = 1.8
+}
```
Comment (ubschmidt2) [inline contrib/zpages/build.gradle]: I'm not too deep into Gradle - just wondering if this is effectively the same as
sourceCompatibility = 1.8
targetCompatibility = 1.8

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An {@code HttpHandler} that displays information about active and sampled spans recorded using
+ * the OpenCensus library.
+ *
+ * <p>Example usage with automatic context propagation:
```
Comment (ubschmidt2) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: nit: This example seems unrelated to (automatic) context propagation.

Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: Done.

File: settings.gradle (hunk 10 above)
Comment (bogdandrutu) [inline settings.gradle]: We are moving everything to opencensus-XXX

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java (hunk 13 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: Was debugging. I will leave some instrumentation because it is nice to have it instrumented :)

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java (hunk 12 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: Done. Added tests.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java (hunk 11 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: Done.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java (hunk 9 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: https://github.com/census-instrumentation/opencensus-java/issues/520

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java (hunk 8 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: :1st_place_medal:

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+      httpExchange.close();
+    }
+  }
+
+  private static Map<String, String> queryToMap(HttpExchange httpExchange) {
+    String query = httpExchange.getRequestURI().getQuery();
+    if (query == null) {
+      return Collections.emptyMap();
+    }
+    Map<String, String> result = new HashMap<String, String>();
```
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezHttpHandler.java]: From URI.getQuery():
Returns the decoded query component of this URI.
The string returned by this method is equal to that returned by the getRawQuery method except that all sequences of escaped octets are decoded.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+                ? 0
+                : durationToNanos(
+                    span.getEndTimestamp().subtractTimestamp(span.getStartTimestamp())))
+            * 1.0e-9;
+
+    formatter.format(
+        "<b>%04d/%02d/%02d-%02d:%02d:%02d.%06d %13.6f     TraceId: %s SpanId: %s "
+            + "ParentSpanId: %s</b>%n",
+        calendar.get(Calendar.YEAR),
+        calendar.get(Calendar.MONTH) + 1,
```
Comment (sebright) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: It would be nice to have tests for the date and time formatting.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+    return new TracezPageFormatter(runningSpanStore, sampledSpanStore);
+  }
+
+  /**
+   * Emits the HTML generated page to the {@code outputStream}.
+   *
+   * @param queryMap the query components map.
+   * @param outputStream the output {@code OutputStream}.
+   */
+  public void emitHtml(Map<String, String> queryMap, OutputStream outputStream) {
```
Comment (sebright) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Would it make sense to use a higher level HTML formatting library or templating, such as https://j2html.com/ or https://developers.google.com/closure/templates/?  I haven't worked with HTML generation in Java, so I don't know what is the best choice for a webpage of this complexity.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 17 above)
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Added a TODO.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+    }
+    stringBuilder.append("}");
+    return stringBuilder.toString();
+  }
+
+  private static String attributeValueToString(AttributeValue attributeValue) {
+    return attributeValue.match(
+        new Function<String, String>() {
+          @Override
+          public String apply(String stringValue) {
```
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Trying to keep the code java7 compatible. I will try to investigate and understand why cannot run on openjdk7 but works with sun-jdk7.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java [hunk 20]
```diff
@@ -0,0 +1,630 @@
...
+      implements Comparator<TimedEvent<?>>, Serializable {
+    private static final long serialVersionUID = 0;
+
+    @Override
+    public int compare(TimedEvent<?> o1, TimedEvent<?> o2) {
+      return o1.getTimestamp().compareTo(o2.getTimestamp());
+    }
+  }
+
+  private static final class SpanDataComparator implements Comparator<SpanData>, Serializable {
```
Comment (bogdandrutu) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: findbugs:
SE_COMPARATOR_SHOULD_BE_SERIALIZABLE: Comparator doesn't implement Serializable

File: contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+  @Test
+  public void linksForActiveRequests_InSummaryTable() {
+    OutputStream output = new ByteArrayOutputStream();
+    TracezPageFormatter tracezPageFormatter =
+        TracezPageFormatter.create(runningSpanStore, sampledSpanStore);
+    when(runningSpanStore.getSummary()).thenReturn(runningSpanStoreSummary);
+    when(sampledSpanStore.getSummary()).thenReturn(sampledSpanStoreSummary);
+    tracezPageFormatter.emitHtml(Collections.emptyMap(), output);
+    // 3 active requests
+    assertThat(output.toString()).contains("href='?zspanname=TestActiveSpan&ztype=0&zsubtype=0'>3");
```
Comment (bogdandrutu) [inline contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java]: we can do something like that.

File: contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+    tracezPageFormatter.emitHtml(Collections.emptyMap(), output);
+    // No sampled links (ztype=1);
+    assertThat(output.toString()).doesNotContain("href=\"?zspanname=TestActiveSpan&ztype=2");
+    // Links for 7 errors 2 CANCELLED + 5 DEADLINE_EXCEEDED;
+    assertThat(output.toString())
+        .contains("href='?zspanname=TestSampledSpan&ztype=2&zsubtype=0'>7");
+    // Links for 7 errors 2 CANCELLED + 5 DEADLINE_EXCEEDED;
+    assertThat(output.toString())
+        .contains("href='?zspanname=TestActiveAndSampledSpan&ztype=2&zsubtype=0'>7");
+  }
```
Comment (bogdandrutu) [inline contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java]: Added TODO.

File: contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+    numbersOfErrorSampledSpans.put(CanonicalCode.DEADLINE_EXCEEDED, 5);
+    Map<String, SampledSpanStore.PerSpanNameSummary> sampledSummaryMap = new HashMap<>();
+    sampledSummaryMap.put(
+        SAMPLED_SPAN_NAME,
+        SampledSpanStore.PerSpanNameSummary.create(
+            numbersOfLatencySampledSpans, numbersOfErrorSampledSpans));
+    sampledSummaryMap.put(
+        ACTIVE_SAMPLED_SPAN_NAME,
+        SampledSpanStore.PerSpanNameSummary.create(
+            numbersOfLatencySampledSpans, numbersOfErrorSampledSpans));
```
Comment (bogdandrutu) [inline contrib/zpages/src/test/java/io/opencensus/zpages/TracezPageFormatterTest.java]: Not necessary, the links have names as well.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 18 above)
Comment (sebright) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Okay.  I wasn't recommending j2html specifically, I just wanted to give an example of what I meant.

File: contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java (hunk 19 above)
Comment (sebright) [inline contrib/zpages/src/main/java/io/opencensus/zpages/TracezPageFormatter.java]: Okay, I didn't realize it was Java 7.
